src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIApp.cs

```diff
@@ -44,27 +63,53 @@ public static class GenAPIApp
 
             string headerFileText = ReadHeaderFile(headerFile);
 
-            AccessibilitySymbolFilter accessibilitySymbolFilter = new(
+            ISymbolFilter typeFilter = new AccessibilitySymbolFilter(
                 respectInternals,
                 includeEffectivelyPrivateSymbols: true,
                 includeExplicitInterfaceImplementationSymbols: true);
 
+            DocIdSymbolFilter? includeAdditionalApiFilter = null;
+
+            if (includeApiFiles is not null)
+            {
+                includeAdditionalApiFilter = new DocIdSymbolFilter(includeApiFiles, includeDocIds: true);
+            }
+
+            if (!excludeInternalCompilerAttributes)
+            {
+                includeAdditionalApiFilter ??= new DocIdSymbolFilter(includeDocIds: true);
+                includeAdditionalApiFilter.DocIds.UnionWith(s_compilerAttributes);
```

Comment:
I think it would be better to combine this so that `DocIds` doesn't need to be exposed and can't be mutated from the outside.

Reply:
The only constructor we have expects file-lists.  We should not create a file just to pass it in to that constructor.  It felt clunky to add more constructor overloads just to accept both file-lists and doc-ids.  It seemed reasonable to expose the list - we do expose mutable members on other filters (`CompositeSymbolFilter.Filters`).

I could instead expose Add/AddRange methods add more types to the docID list.  Would that be preferrable, or do you feel strongly about this filter (and not CompositeSymbolFilter) being immutable?

Reply:
`CompositeSymbolFilter.Filters` should be read-only. Please change the return type to something read-only instead of `List<T>`. That's just an oversight on our side.

I don't feel super strongly about filters being immutable but I would prefer if they were...

Reply:
Well all the usage of `CompsiteFilter.Filters` already mutates the filter list - and there is no other way to put data in that list. It's clearly designed to be mutable 😕

Reply:
Somewhat Ironically, making `CompositeSymbolFilter.Filters` an `IReadOnlyList<T>` removes the ability to add values to it with the collection initializer - forcing me to use the `Add` method.

Reply:
I tried simplifying this locally and pushed a branch to my repo so that you can take a look (I didn't want to directly push into your PR): https://github.com/dotnet/sdk/compare/main...ViktorHofer:sdk:GenAPIInternalAttributesPR?expand=1

Reply:
Thank you for sharing your suggestion - I was thinking about this but didn't know what direction you'd prefer.   It seems you kept CompositeTypeFilter as mutable but made DocIdSymbolFilter immutable.   Could you help me understand this preference?

I was intentionally trying to avoid two DocIdFilters as the DocID calculation is [not a cached by roslyn](https://github.com/dotnet/roslyn/blob/ed20cf0fc50854b65da285bff9666acbd4b4dfc2/src/Compilers/CSharp/Portable/Symbols/Symbol.cs#L982-L984) and we'll be doing this for every single symbol we encounter - which we also do for the exclusion filter.   It could be premature optimization - I could measure it if you think it's important.  

I'll defer to you on your preference for the API shape of these filters.

Reply:
I applied your changes with a small tweak to specify the mode enum rather than rely on default.

Reply:
Here's benchmark results comparing traversal of the symbols in System.Private.CoreLib with 1 DocIdFilter vs 2.

| Method               | Mean     | Error     | StdDev    |
|--------------------- |---------:|----------:|----------:|
| OneDocIdSymbolFilter | 3.409 ms | 0.0648 ms | 0.0575 ms |
| TwoDocIdSymbolFilter | 3.698 ms | 0.0692 ms | 0.1193 ms |

So while it does make a measurable difference, perhaps this isn't on the scale to matter for the overall performance of the tool.